packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,478 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, unreachableCase } from "@fluidframework/common-utils";
+import {
+	appendToMergeTreeDeltaRevertibles as appendToMergeTreeRevertibles,
+	IMergeTreeDeltaCallbackArgs,
+	LocalReferencePosition,
+	PropertySet,
+	MergeTreeDeltaRevertible,
+	ReferenceType,
+	MergeTreeDeltaType,
+	refTypeIncludesFlag,
+	revertMergeTreeDeltaRemoveRevertible,
+	revertMergeTreeDeltaRevertibles,
+	ISegment,
+} from "@fluidframework/merge-tree";
+import { IntervalCollection, SequenceInterval } from "./intervalCollection";
+import { SharedString, SharedStringSegment } from "./sharedString";
+
+export type SharedStringRevertible = MergeTreeDeltaRevertible | IntervalRevertible;
+
+export const IntervalEventType = {
+	CHANGE: 0,
+	ADD: 1,
+	DELETE: 2,
+	PROPERTYCHANGED: 3,
+	SEQUENCEREMOVE: 4,
+} as const;
+
+type IntervalEventType = typeof IntervalEventType[keyof typeof IntervalEventType];
+
+export type IntervalRevertible =
+	| {
+			event: typeof IntervalEventType.CHANGE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+	  }
+	| {
+			event: typeof IntervalEventType.ADD;
+			interval: SequenceInterval;
+	  }
+	| {
+			event: typeof IntervalEventType.DELETE;
+			interval: SequenceInterval;
+			start: LocalReferencePosition;
+			end: LocalReferencePosition;
+	  }
+	| {
+			event: typeof IntervalEventType.PROPERTYCHANGED;
+			interval: SequenceInterval;
+			propertyDeltas: PropertySet;
+	  }
+	| {
+			event: typeof IntervalEventType.SEQUENCEREMOVE;
+			intervals: {
+				intervalId: string;
+				label: string;
+				startSegmentIndex?: number;
```

Comment:
it looks like these are set using the offset of the ref at append time. Haven't read through the whole PR yet, but I think you don't need to store this explicitly because you can use `start.getOffset()`

Reply:
oh wait I was reading it incorrectly, it's the index of the segment in the delta event.

Reply:
i'm also still reading, but i don't get how this can be stable. segments can still split after remove